README.md
# nosoup-clj v0.1.2

Site generator for https://www.nosoupforyou.com/, a listing of local &
independent restaurants in Gainesville, FL I created in the early
2000's. The original project was written with PHP/MySQL and over the
past 8 or so years, it'd been running on AWS. I decided to move away
from all that and rewrote the whole thing as a Clojure app that
generates static pages. Instead of a database, I'm using a simple EDN
configuration.

## Usage

I can't imagine anyone will want to do this but, assuming you've
created your own configuration, you generate the static pages with:

    $ java -jar nosoup-clj-0.1.2-standalone.jar gen path-to-restaurants-config.edn

See [test/restaurants.edn](test/restaurants.edn) for an example.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.1.2] - 2020-05-15
### Added
- HTML cleanup fn.

### Changed
- rework map links
- redo UI a bit. Switch to new & simpler site.css. Cleanup.

## [0.1.1] - 2020-05-04
### Changed
- Use localeCompare in JS.
- CSS cleanup and partial minification.
- Tests now use hickory for easier check of output.
- Moved test resources to test/ directory.

### Added
- More meta details in each output page head.
- Generate sitemap.xml
- Full spec validation. Implemented generators.
- Don't rewrite unchanged content to prevent timestamp modification.

### Fixed
- Corrected / improved all tests.
- Delete stale HTML output from categories that no longer have defined restaurants.

## [0.1.0] - 2020-04-20
- Initial release with most features from original site.

[0.1.2]: https://github.com/edporras/nosoup-clj/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/edporras/nosoup-clj/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/edporras/nosoup-clj/releases/tag/v0.1.0


test/nosoup_clj/core_test.clj
(ns nosoup-clj.core-test
  (:require [clojure.java.io :as io]
            [clojure.string  :as str]
            [clojure.spec.alpha     :as s]
            [clojure.spec.gen.alpha :as gen]
            [clojure.test    :refer [deftest is testing]]
            [ring.util.codec :as codec :refer-only [form-encode]]
            [hickory.core    :as html]
            [nosoup-clj.core :as sut]
            [nosoup-clj.spec :as spec]))

(defn- html->hiccup
  [html-str]
  (->> html-str
       (html/parse-fragment)
       (mapv html/as-hiccup)))

(def test-categories (io/file "test/categories.edn"))
(deftest read-categories-list-adds-all-entry-test
  (testing "Entry for `:all` is added to read categories."
    (is (= "All" (:all (sut/read-categories-list test-categories))))))

(def test-restaurants (io/file "test/restaurants.edn"))
(deftest read-restaurant-list-filters-closed-test
  (testing "Restaurants marked `:closed` are filtered."
    (let [restaurants (sut/read-restaurant-list test-restaurants)]
      (is (empty? (->> restaurants
                       (filter #(contains? (:opts %) :closed))))))))

(deftest categories->html-selected-test
  (testing "Category options list selects the correct option."
    (is (= (sut/categories->html {:a "A" :mexican "Mexican" :c "C"} :mexican)
           [:nav
            [:form
             {:name "catlist", :method :get, :action "/c"}
             [:select
              {:name "cat", :size 1, :onchange "selChange();"}
              '([:option {:value "a"} "A"]
                [:option {:value "mexican", :selected "selected"} "Mexican"]
                [:option {:value "c"} "C"])]
             [:input
              {:type "submit", :name "action", :id "search", :value "Search"}]]]))))

(deftest link-data->html-test
  (testing "Generate HTML output for a uri text tuple with the correct attributes."
    (let [output (sut/link-data->html ["/abc/" "ABC"])]
      (is (= (html->hiccup output)
             [[:a {:href "/abc/"} "ABC"]])))))

(deftest link-data->html-with-external-link-test
  (testing "Generate HTML output for a uri text tuple with the correct attributes for external links."
    (let [output (sut/link-data->html ["http://abc.de/" "ABC"])]
      (is (= (html->hiccup output)
             [[:a
               {:href "http://abc.de/",
                :rel "noopener noreferrer",
                :target "_blank"}
               "ABC"]])))))

(deftest restaurant-links-single-entry-test
  (testing "Generation of list of links with a single entry (separator not used)."
    (is (= (sut/restaurant-links "Test: " " * " [["/abc/" "ABC"]])
           (list "Test: " "<a href=\"/abc/\">ABC</a>")))))

(deftest restaurant-links-multiple-entries-test
  (testing "Generation of list of links with multiple entries should include separator."
    (is (= (sut/restaurant-links "Test: " " + " [["/abc/" "ABC"] ["/def/" "DEF"] ["/ghi/" "GHI"]])
           (list "Test: "
                 "<a href=\"/abc/\">ABC</a> + <a href=\"/def/\">DEF</a> + <a href=\"/ghi/\">GHI</a>")))))

(deftest restaurant-links-nil-url-dropped-test
  (testing "Generation of restaurant links with `nil` url value is dropped."
    (is (= (sut/restaurant-links "Test: " " / " [["/abc/" "ABC"] [nil "DEF"] ["/ghi/" "GHI"]])
           (list "Test: " "<a href=\"/abc/\">ABC</a> / <a href=\"/ghi/\">GHI</a>")))))

(deftest restaurant-links-nil-entry-dropped-test
  (testing "Generation of restaurant links with `nil` entry is dropped."
    (is (= (sut/restaurant-links "Test: " " = " [["/abc/" "ABC"] nil ["/ghi/" "GHI"]])
           (list "Test: " "<a href=\"/abc/\">ABC</a> = <a href=\"/ghi/\">GHI</a>")))))

(deftest restaurant-map-link-url-test
  (testing "Generate map link with form-encoded text."
    (is (= (sut/restaurant-map-link-url {:name "A B C's" :address "123 Main Rd." :city "Towns ville"})
           (str sut/base-mapping-url "A+B+C%27s,Towns+ville,FL")))))

(deftest restaurant-map-link-url-with-coords-test
  (testing "Generate map link including address when `:coords` present."
    (let [rest   (assoc (gen/generate (s/gen ::spec/restaurant)) :coords (gen/generate (s/gen :restaurant/coords)))
          markup (sut/restaurant-map-link-url rest)]
      (is (str/includes? markup (codec/form-encode (:address rest)))))))

(deftest twitter-link-data-test
  (testing "Generate twitter link data."
    (is (= (sut/twitter-link-data "test")
           [(str sut/base-twitter-url "test") "@test"]))))

(deftest twitter-link-data-with-nil-test
  (testing "Generate twitter link data."
    (is (nil? (sut/twitter-link-data nil)))))

(deftest restaurant-category-listing-single-all-test
  (testing "Category list with single entry, `:all` selected, creates link output with 'Under'."
    (let [rest       {:name "ABC" :address "" :city "" :zip "" :phone "" :categories #{:mexican}}
          categories {:a "A" :b "B" :c "C" :italian "Italian" :mexican "Mexican"}]
      (is (= (sut/restaurant-category-listing rest categories :all)
             (list "Under: " "<a href=\"/mexican/\">Mexican</a>"))))))

(deftest restaurant-category-listing-single-selected-test
  (testing "Category list with single entry, same selected, generates &nbsp; for spacing."
    (let [rest       {:name "ABC" :address "" :city "" :zip "" :phone "" :categories #{:mexican}}
          categories {:a "A" :b "B" :c "C" :italian "Italian" :mexican "Mexican"}]
      (is (= "&nbsp" (sut/restaurant-category-listing rest categories :mexican))))))

(deftest restaurant-category-listing-multiple-other-selected-test
  (testing "Category list with multiple entries, one selected, creates link output with 'Also under'."
    (let [rest       {:name "ABC" :address "" :city "" :zip "" :phone "" :categories #{:mexican :italian}}
          categories {:a "A" :b "B" :c "C" :italian "Italian" :mexican "Mexican"}]
      (is (= (sut/restaurant-category-listing rest categories :mexican)
             (list "Also under: " "<a href=\"/italian/\">Italian</a>"))))))

(deftest restaurant-by-category-single-category-test
  (testing "Restaurant list is filtered by the given category."
    (let [restaurants [{:name "Restaurant 1" :address "A" :city "A" :zip "12345" :phone "123 456-7890" :categories #{:italian}}
                       {:name "Restaurant 2" :address "B" :city "B" :zip "12345" :phone "123 456-7890" :categories #{:mexican}}
                       {:name "Restaurant 3" :address "C" :city "C" :zip "12345" :phone "123 456-7890" :categories #{:american}}]]
      (is (= (sut/restaurant-by-category restaurants :mexican)
             [{:name "Restaurant 2" :address "B" :city "B" :zip "12345" :phone "123 456-7890" :categories #{:mexican}}])))))

(deftest restaurant-by-category-all-test
  (testing "Restaurant list is not filtered when given `:all`"
    (let [restaurants (gen/generate (s/gen ::spec/restaurants))]
      (is (= (sut/restaurant-by-category restaurants :all)
             restaurants)))))

(deftest restaurants->html-single-category-test
  (testing "Restaurant listing output with single category."
    (let [rest-list [{:name "Rest Name" :address "B" :city "C" :zip "22222" :phone "123 456-7890" :categories #{:mexican}}]
          output    (sut/restaurants->html rest-list {:mexican "Mexican"} :mexican)]
      (is (= (html->hiccup output)
             [[:li
               {}
               [:h2 {} "Rest Name"]
               [:div
                {:class "info"}
                [:a {:href "tel:+1-123-456-7890"} "123 456-7890"]
                [:address {} 
                 [:a {:href (str sut/base-mapping-url "Rest+Name,C,FL")
                      :rel "noopener noreferrer"
                      :target "_blank"}
                  "B" [:br {}] "C, FL 22222"]]
                [:div {:class "links"}]]
               [:footer {} " "]]])))))

(deftest restaurants->html-single-category-with-alias-test
  (testing "Restaurant listing output with single category and an alias uses alias in heading."
    (let [rest-list [{:name "Rest Name" :alias "Resty" :address "B" :city "C" :zip "22222" :phone "123 456-7890" :categories #{:mexican}}]
          output    (sut/restaurants->html rest-list {:mexican "Mexican"} :mexican)
          h2-text   (-> (map html/as-hickory (html/parse-fragment output))
                        first
                        (get-in [:content 0 :content]))]
      (is (= h2-text
             ["Resty"])))))

(deftest restaurants->html-single-category-with-twitter-test
  (testing "Restaurant listing output with single category and twitter link."
    (let [rest-list [{:name "Rest Name" :address "B" :city "C" :zip "22222" :phone "123 456-7890" :twitter "rn" :categories #{:mexican}}]
          output    (sut/restaurants->html rest-list {:mexican "Mexican"} :mexican)
          twitter   (-> (map html/as-hickory (html/parse-fragment output))
                        first
                        (get-in [:content 1 :content 2 :content 1])
                        (dissoc :type :tag))]
      (is (= twitter
             {:attrs {:href (str sut/base-twitter-url "rn"), :rel "noopener noreferrer", :target "_blank"},
              :content ["@rn"]})))))

(deftest generate-category-page-head-includes-title-test
  (testing "Generating head portion of page includes title tag."
    (let [output (sut/generate-category-page-head [:all "All"])]
      (is (= (->> output
                  (filter #(and (vector? %) (= :title (first %))))
                  count)
             1)))))

(deftest generate-category-page-head-includes-meta-author-test
  (testing "Generating head portion of page includes author meta info."
    (let [output (sut/generate-category-page-head [:all "All"])]
      (is (= (->> output
                  (filter #(and (vector? %) (= :meta (first %))))
                  (filter (fn [[_ v]] (= "author" (:name v))))
                  count)
             1)))))

(deftest generate-category-page-head-adds-category-to-keywords-test
  (testing "Generating head portion of page when the category is set appends it to keyword meta."
    (let [category [:italian "Italian"]
          output   (sut/generate-category-page-head category)
          keywords (->> output
                        (filter #(and (vector? %) (= :meta (first %))))
                        (filter (fn [[_ v]] (= "keywords" (:name v))))
                        (map (fn [[_ v]] (:content v)))
                        first)]
      (is (str/includes? keywords (last category))))))

(deftest generate-category-page-head-adds-category-to-description-test
  (testing "Generating head portion of page when the category is set appends it to description meta."
    (let [category [:italian "Italian"]
          output   (sut/generate-category-page-head category)
          keywords (->> output
                        (filter #(and (vector? %) (= :meta (first %))))
                        (filter (fn [[_ v]] (= "description" (:name v))))
                        (map (fn [[_ v]] (:content v)))
                        first)]
      (is (str/includes? keywords (last category))))))

(deftest generate-category-page-sets-lang-test
  (testing "Generated HTML for a page has language set to 'en'."
    (let [output  (sut/generate-category-page [:italian "Italian"]
                                              [{:name "Test 1" :address "Address 1" :city "City" :zip "12345" :phone "123 456-7890" :categories #{:italian}}]
                                              {:italian "Italian"}
                                              {:italian "Italian"})
          content (->> (:content (html/as-hickory (html/parse output)))
                       (filter #(= :html (:tag %)))
                       first)]
      (is (= (get-in content [:attrs :lang])
             "en")))))

(deftest generate-category-page-test
  (testing "Generated HTML from category + restaurant data with selected category."
    (is (= (sut/generate-category-page [:italian "Italian"]
                                       [{:name "Test 1" :address "Address 1" :city "City" :zip "12345" :phone "123 456-7890" :categories #{:italian}}]
                                       {:italian "Italian"}
                                       {:italian "Italian"})
           "<!DOCTYPE html>\n<html lang=\"en\"><head><title>No Soup For You - Gainesville - Italian</title><meta content=\"Ed Porras\" name=\"author\"><meta content=\"Guide of independent restaurants and grocers in Gainesville, FL under the Italian category\" name=\"description\"><meta content=\"Gainesville Local Independently-owned Restaurants Italian\" name=\"keywords\"><meta content=\"width=device-width,initial-scale=1.0\" name=\"viewport\"><link href=\"/css/site.css\" rel=\"stylesheet\" type=\"text/css\"><script src=\"/js/site.js\" type=\"text/javascript\"></script></head><body onload=\"load();\"><header><h1><img alt=\"Dining in Gainesville\" height=\"42\" src=\"/img/logo.png\" width=\"293\"></h1><p>Locally-owned restaurants, cafes, and grocers.</p><nav><form action=\"/c\" method=\"get\" name=\"catlist\"><select name=\"cat\" onchange=\"selChange();\" size=\"1\"><option selected=\"selected\" value=\"italian\">Italian</option></select><input id=\"search\" name=\"action\" type=\"submit\" value=\"Search\"></form></nav></header><div id=\"content\"><ul><li><h2>Test 1</h2><div class=\"info\"><a href=\"tel:+1-123-456-7890\">123 456-7890</a><address><a href=\"https://maps.google.com/?daddr=Test+1,City,FL\" rel=\"noopener noreferrer\" target=\"_blank\">Address 1<br />City, FL 12345</a></address><div class=\"links\"></div></div><footer>&nbsp</footer></li></ul></div><footer><p>This is a listing of independent businesses in Gainesville, FL. If you own or know of a business you'd like to see listed, please contact: nsfy at digressed dot net or via Twitter at <a href=\"https://twitter.com/NSFYgnv\" rel=\"noopener noreferrer\" target=\"_blank\">@NSFYgnv</a>.</p></footer></body></html>"))))

(deftest generate-category-page-for-all-test
  (testing "Generating HTML from category + restaurant data with category `:all`."
    (is (= (sut/generate-category-page [:all "All"]
                                       [{:name "Test 1" :address "Address 1" :city "City" :zip "12345" :phone "123 456-7890" :categories #{:italian}}]
                                       {:italian "Italian"}
                                       {:italian "Italian"})
           "<!DOCTYPE html>\n<html lang=\"en\"><head><title>No Soup For You - Gainesville</title><meta content=\"Ed Porras\" name=\"author\"><meta content=\"Guide of independent restaurants and grocers in Gainesville, FL\" name=\"description\"><meta content=\"Gainesville Local Independently-owned Restaurants\" name=\"keywords\"><meta content=\"width=device-width,initial-scale=1.0\" name=\"viewport\"><link href=\"/css/site.css\" rel=\"stylesheet\" type=\"text/css\"><script src=\"/js/site.js\" type=\"text/javascript\"></script></head><body onload=\"load();\"><header><h1><img alt=\"Dining in Gainesville\" height=\"42\" src=\"/img/logo.png\" width=\"293\"></h1><p>Locally-owned restaurants, cafes, and grocers.</p><nav><form action=\"/c\" method=\"get\" name=\"catlist\"><select name=\"cat\" onchange=\"selChange();\" size=\"1\"><option value=\"italian\">Italian</option></select><input id=\"search\" name=\"action\" type=\"submit\" value=\"Search\"></form></nav></header><div id=\"content\"><ul><li><h2>Test 1</h2><div class=\"info\"><a href=\"tel:+1-123-456-7890\">123 456-7890</a><address><a href=\"https://maps.google.com/?daddr=Test+1,City,FL\" rel=\"noopener noreferrer\" target=\"_blank\">Address 1<br />City, FL 12345</a></address><div class=\"links\"></div></div><footer>Under: <a href=\"/italian/\">Italian</a></footer></li></ul></div><footer><p>This is a listing of independent businesses in Gainesville, FL. If you own or know of a business you'd like to see listed, please contact: nsfy at digressed dot net or via Twitter at <a href=\"https://twitter.com/NSFYgnv\" rel=\"noopener noreferrer\" target=\"_blank\">@NSFYgnv</a>.</p></footer></body></html>"))))

(deftest generate-category-restaurant-list-test
  (testing "Generate category to restaurant list map."
    (is (= (sut/generate-category-restaurant-list {:italian "Italian" :mexican "Mexican"}
                                                  (sut/read-restaurant-list test-restaurants))
           [[:italian '({:name "Test 1" :address "Address 1" :city "City" :zip "12345" :phone "123 456-7890" :categories #{:italian}})]
            [:mexican '({:name "Test 3" :address "Address 3" :city "City" :zip "12345" :phone "123 456-7892" :categories #{:mexican}}
                        {:name "Test 4" :address "Address 4" :city "City" :zip "12345" :phone "123 456-7893" :categories #{:mexican}})]]))))

(deftest generate-category-restaurant-list-empty-test
  (testing "Generate category to restaurant list map produces entry w/ empty list when none are found."
    (is (= (sut/generate-category-restaurant-list {:american "American"}
                                                  (sut/read-restaurant-list test-restaurants))
           [[:american '()]]))))

(deftest filter-category-list-from-generated-restaurant-data-test
  (testing  "Removes entries from the category list that didn't generate any page output."
    (is (= (sut/filter-category-list-from-generated-restaurant-data [[:all '([])] [:italian '([])]] {:all "A" :italian "I" :mexican "M" :latin "L"})
           {:all "A" :italian "I"}))))

(deftest filter-category-list-from-generated-restaurant-data-returns-sorted-list-test
  (testing  "Filtered category list is sorted."
    (let [output (sut/filter-category-list-from-generated-restaurant-data [[:italian '([])] [:all '([])] [:vietnamese '([])]] {:all "A" :italian "I" :mexican "M" :latin "L" :chinese "C"})]
      (is (= output
             (into (sorted-map) output))))))

(deftest category-page-output-path-for-root-test
  (testing "Output path for main index page.")
  (is (= (sut/category-page-output-path "a-path/" :all)
         "a-path/index.html")))

(deftest category-page-output-path-for-subcategory-test
  (testing "Output path for sub-category index page includes subpath.")
  (is (= (sut/category-page-output-path "a-path/" :italian)
         "a-path/italian/index.html")))


test/nosoup_clj/init_test.clj
(ns nosoup-clj.init-test
  (:require [clojure.test :refer [deftest is]]
            [nosoup-clj.init :as sut]))

(deftest validate-args-test
  (let [status (sut/validate-args '("gen" "test/restaurants.edn"))]
    (is (and (= (:action status) "gen")
             (= (get-in status [:options :config]) "test/restaurants.edn")
             (not (nil? (get-in status [:options :base-output-path])))))))

(deftest validate-args-with-basepath-opt-test
  (is (= (-> (sut/validate-args '("gen" "test/restaurants.edn" "-o" "test/"))
             (get-in [:options :base-output-path]))
         "test/")))

(deftest validate-args-no-file-error-test
  (is (= (keys (sut/validate-args '("gen")))
         '(:exit-message))))

(deftest validate-args-help-option-test
  (let [status (sut/validate-args '("gen" "-h"))]
    (is (and (= (keys status) '(:exit-message :ok?))
             (:ok? status)))))

(deftest validate-args-file-error-test
  (is (= (keys (sut/validate-args '("gen" "blah.edn")))
         '(:exit-message))))

(deftest validate-args-unknown-command-test
  (is (= (keys (sut/validate-args '("what")))
         '(:exit-message))))


test/nosoup_clj/spec_test.clj
(ns nosoup-clj.spec-test
  (:require [clojure.test                  :refer [deftest is]]
            [clojure.spec.alpha            :as s]
            [clojure.test.check.generators :as gen]
            [nosoup-clj.spec               :as sut]))

(deftest check-spec-returns-data-test
  (let [categories {:italian "Italian" :chinese "Chinese"}]
    (is (= (sut/check-spec ::sut/categories categories)
           categories))))

(deftest zip-spec-not-string-test
  (is (not= :s/invalid (s/conform :restaurant/zip 32321))))

(deftest phone-spec-not-string-test
  (is (not= :s/invalid (s/conform :restaurant/phone 3521234567))))

(deftest phone-spec-not-formatted-test
  (is (not= :s/invalid (s/conform :restaurant/phone "352-123-4567"))))

(deftest uri-spec-not-string-test
  (is (not= :s/invalid (s/conform :restaurant/uri :temp))))

(deftest dine-opts-spec-not-set-test
  (is (not= :s/invalid (s/conform :restaurant/dine-opts [:delivery]))))

(deftest dine-opts-spec-unknown-test
  (is (not= :s/invalid (s/conform :restaurant/dine-opts #{:what-is-this}))))

(deftest opts-spec-not-set-test
  (is (not= :s/invalid (s/conform :restaurant/opts [:wifi]))))

(deftest restaurant-uri-gen-test
  (is (s/valid? :restaurant/uri (gen/generate (s/gen :restaurant/uri)))))

(deftest restaurant-dine-opts-gen-test
  (is (s/valid? :restaurant/dine-opts (gen/generate (s/gen :restaurant/dine-opts)))))

(deftest restaurant-price-gen-test
  (is (s/valid? :restaurant/price (gen/generate (s/gen :restaurant/price)))))

(deftest restaurant-coords-gen-test
  (is (s/valid? :restaurant/coords (gen/generate (s/gen :restaurant/coords)))))

(deftest restaurant-gen-test
  (is (s/valid? ::sut/restaurant (gen/generate (s/gen ::sut/restaurant)))))


project.clj
(defproject nosoup-clj "0.1.2"
  :description "Site generator for nosoupforyou.com"
  :url "https://github.com/edporras/nosoup-clj/"
  :license {:name "MIT"
            :url "https://opensource.org/licenses/mit-license.php"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.clojure/tools.cli "1.0.194"]

                 [hiccup "1.0.5"]
                 [ring/ring-codec "1.1.2"]
                 [clojure.java-time "0.3.2"]
                 [sitemap "0.4.0"]
                 [digest "1.4.9"]

                 [com.taoensso/timbre "4.10.0"]]
  :main ^:skip-aot nosoup-clj.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}
             :test {:dependencies [[hickory "0.7.1"]
                                   [org.clojure/test.check "1.0.0"]
                                   [com.oscaro/tools-io "0.3.19"]]}
             :dev  {:dependencies [[hickory "0.7.1"]
                                   [org.clojure/test.check "1.0.0"]
                                   [com.oscaro/tools-io "0.3.19"]]}
             :repl {:dependencies [[hickory "0.7.1"]
                                   [org.clojure/test.check "1.0.0"]
                                   [com.oscaro/tools-io "0.3.19"]]}})


src/nosoup_clj/core.clj
(ns nosoup-clj.core
  (:require [nosoup-clj.init         :as init]
            [nosoup-clj.util         :as util]
            [nosoup-clj.spec         :as spec]
            [clojure.spec.alpha      :as s]
            [clojure.java.io         :as io]
            [clojure.string          :as str]
            [hiccup.core             :refer [html]]
            [hiccup.page             :as page]
            [ring.util.codec         :as codec :refer-only [form-encode]]
            [taoensso.timbre         :as timbre :refer [info fatal]])
  (:import [java.util Locale]
           [java.text Collator])
  (:gen-class))

(def base-title (str "No Soup For You - Gainesville"))

;; turns out this falls-back to google if the device doesn't support apple maps
(def base-mapping-url "https://maps.google.com/?daddr=")
(def base-twitter-url "https://twitter.com/")

(defn read-categories-list
  "Reads the categories config, injects a global entry used to list `:all`."
  [categories-edn]
  (->> (util/read-config categories-edn)
       (spec/check-spec ::spec/categories)
       (into {:all "All"})))

(def str-collator (Collator/getInstance (Locale. "en_US")))
(defn read-restaurant-list
  "Reads the restaurants config sorting by the name. Filters out any
  restaurants marked `:closed`."
  [restaurants-edn]
  (->> (util/read-config restaurants-edn)
       (spec/check-spec ::spec/restaurants)
       (remove #(contains? (:opts %) :closed))
       (sort-by :name str-collator)))

(defn categories->html
  "Generate the navigation pulldown selector from the category list."
  [categories selected-category]
  {:pre [(s/valid? ::spec/categories categories) keyword? selected-category]}
  [:nav [:form {:name "catlist" :method :get :action "/c"}
         [:select {:name "cat" :size 1 :onchange "selChange();"}
          (for [[k category-str] categories]
            (let [selected (if (= selected-category k)
                             {:selected "selected"}
                             {})]
              [:option (merge {:value (name k)} selected) category-str]))]
         [:input {:type "submit" :name "action" :id "search" :value "Search"}]]])

(defn link-data->html
  "Generate HTML output from a [uri text] tuple."
  [[uri text]]
  {:pre [string? uri string? text]}
  (let [opts (conj {:href uri}
                   (when (str/starts-with? uri "http")
                     {:target "_blank" :rel "noopener noreferrer"}))]
    (html [:a opts text])))

(defn restaurant-links
  "Using a collection of vectors of `[uri text]` entries, generate their
  string output separated by `separator` if there is more than one."
  [label separator link-data]
  {:pre [string? separator vector? link-data]}
  (let [filtered-link-data (remove #(or (nil? %) (nil? (first %))) link-data)]
    (when (seq filtered-link-data)
      (list label (->> (for [link filtered-link-data]
                             (link-data->html link))
                           vec
                           (str/join separator))))))

(defn restaurant-map-link-url
  "Generate the map urk for a restaurant."
  [{:keys [name address city coords]}]
  {:pre [string? name string? city (s/nilable (s/valid? :restaurant/coords coords))]}
  (str base-mapping-url
       (->> [name (when coords address) city "FL"]
              (remove nil?)
              (vec)
              (map #(codec/form-encode %))
              (str/join ","))))

(defn twitter-link-data
  "Generate the map link tuple for the twitter link, to be passed to `restaurant-links`."
  [handle]
  {:pre [(s/nilable (s/valid? :restaurant/twitter handle))]}
  (when handle
    [(str base-twitter-url handle) (str "@" handle)]))

(defn restaurant-category-listing
  "Generate the list of categories to be shown in the `Under`
  label. Show all if in the `:all` view; otherwise, show other not
  matching the selected one."
  [{:keys [categories]} full-categories-list selected-category-key]
  {:pre [set? categories (s/valid? ::spec/categories full-categories-list) keyword? selected-category-key]}
  (let [[label restaurant-categories] (if (= :all selected-category-key)
                                        ["Under: "      categories]
                                        ["Also under: " (remove #{selected-category-key} categories)])]
    (if (seq restaurant-categories)
      (->> restaurant-categories
           (mapv #(let [category-str (get full-categories-list %)]
                    (vec [(str "/" (name %) "/") category-str])))
           (sort)
           (restaurant-links label ", "))
      "&nbsp")))

(defn restaurants->html
  "Generate page output for a restaurant listing based on the selected category."
  [restaurants full-category-list selected-category-key]
  {:pre [(s/valid? ::spec/restaurants restaurants) (s/valid? ::spec/categories full-category-list) keyword? selected-category-key]}
  (html (for [{:keys [name alias address city zip phone uri twitter] :as r} restaurants]
          (let [restaurant-name (or alias name)
                map-link        [(restaurant-map-link-url r)
                                 (html address [:br] city ", FL " zip)]]
            [:li
             [:h2 restaurant-name]
             [:div {:class "info"}
              [:a {:href (str "tel:+1-" (str/replace phone #" " "-"))} phone]
              [:address (link-data->html map-link)]
              [:div {:class "links"}
               (restaurant-links "Links: "
                                 " | "
                                 [[uri "website"]             ;; restaurant's website
                                  (twitter-link-data twitter) ;; twitter link
                                  ])]]
             [:footer (restaurant-category-listing r full-category-list selected-category-key)]]))))

(defn generate-category-page-head
  "Generate the head portion of the page."
  [[category-k category-str :as c]]
  {:pre [(s/valid? ::spec/category c)]}
  [:head [:title (if (= :all category-k)
                   base-title
                   (str base-title " - " category-str))]
   [:meta {:name "author" :content "Ed Porras"}]
   [:meta {:name "description"
           :content (str "Guide of independent restaurants and grocers in Gainesville, FL"
                         (when-not (= :all category-k)
                           (str " under the " category-str " category")))}]
   [:meta {:name "keywords"
           :content (str "Gainesville Local Independently-owned Restaurants"
                         (when-not (= :all category-k)
                           (str " " category-str)))}]
   [:meta {:name "viewport" :content "width=device-width,initial-scale=1.0"}]
   (page/include-css "/css/site.css")
   (page/include-js "/js/site.js")])

(defn generate-category-page
  "Generate page output for the selected category using the filtered list of restaurants and categories."
  [[category-k :as category] filtered-restaurants filtered-category-list full-category-list]
  (page/html5 {:lang "en"}
              (generate-category-page-head category)
              [:body {:onload "load();"}
               [:header
                [:h1 [:img {:src "/img/logo.png" :alt "Dining in Gainesville" :width "293" :height "42"}]]
                [:p "Locally-owned restaurants, cafes, and grocers."]
                (categories->html filtered-category-list category-k)]
               [:div {:id "content"}
                [:ul
                 (restaurants->html filtered-restaurants full-category-list category-k)]]
               [:footer [:p
                         "This is a listing of independent businesses in Gainesville, FL. If you own or know "
                         "of a business you'd like to see listed, please contact: nsfy at digressed dot net or "
                         "via Twitter at " (link-data->html (twitter-link-data "NSFYgnv")) "."]]]))

(defn restaurant-by-category
  "Filter the list of restaurants using the given `selected-category`."
  [restaurants selected-category]
  {:pre [(s/valid? ::spec/restaurants restaurants) keyword? selected-category]}
  (if (= selected-category :all)
    restaurants ;; don't filter anything
    (->> restaurants
         (filter #(contains? (:categories %) selected-category)))))

(defn generate-category-restaurant-list
  "Create a vector of '(category filtered-restaurant-data) entries."
  [categories restaurants]
  (->> categories
       (mapv (fn [[category-k _]]
               [category-k (restaurant-by-category restaurants category-k)]))))

(defn filter-category-list-from-generated-restaurant-data
  "Removes entries from the category list that didn't generate any page output."
  [category-rest-data full-category-list]
  (->> category-rest-data
       (map (fn [[k data]] (when (seq data) k)))
       (remove nil?)
       (select-keys full-category-list)
       (into (sorted-map))))

(defn category-page-output-path
  [base-output-path category-k]
  (str base-output-path (when (not= :all category-k) (str (name category-k) "/")) "index.html"))

(defn generate-site
  [{:keys [full-restaurant-list full-category-list base-output-path]}]
  {:pre [string? base-output-path]}
  (info (str "read " (count full-category-list) " categories and " (count full-restaurant-list) " restaurants"))
  (let [category-rest-data      (generate-category-restaurant-list full-category-list full-restaurant-list)
        filtered-category-list  (filter-category-list-from-generated-restaurant-data category-rest-data full-category-list)]
    (doseq [[category-k filtered-restaurants] category-rest-data]
      (let [output-path (category-page-output-path base-output-path category-k)]
        (if (seq filtered-restaurants)
          (let [category-str (category-k full-category-list)]
            (info (str "generating output for '" category-k "' with " (count filtered-restaurants) " entries..."))
            (->> (generate-category-page [category-k category-str] filtered-restaurants filtered-category-list full-category-list)
                 (util/cleanup-markup)
                 (util/output->disk output-path)))
          (when (.exists (io/file output-path))
            (info (str "No restaurants found for category " category-k " - deleting old output at " output-path))
            (io/delete-file output-path)
            (io/delete-file (str/replace output-path #"index.html" "")))
          )))
    (util/generate-sitemap base-output-path filtered-category-list)))

(defn -main
  [& args]
  (let [{:keys [action options exit-message ok?]} (init/validate-args args)]
    (if-not exit-message
      (case action
        "gen" (let [full-categories-list {:full-category-list (read-categories-list init/categories-config)}
                    full-restaurant-list {:full-restaurant-list (read-restaurant-list (:config options))}]
                (generate-site (merge options full-restaurant-list full-categories-list))))
      (do ;; error validating args
        (fatal exit-message)
        (System/exit (if ok? 0 1))))))

(comment

  (def categories  (read-categories-list init/categories-config))
  (def restaurants (read-restaurant-list (io/file "test/restaurants.edn")))

  (generate-site {})
  )
